echopype/clean/utils.py

```diff
@@ -0,0 +1,228 @@
+                            channel=channel_index,
+                            range_sample=range_sample_index,
+                            ping_time=ping_time_index,
+                        )
+                    ] = aggregate_window_Sv_value
+
+    return pooled_Sv
+
+
+def _upsample_using_mapping(downsampled_Sv: np.ndarray, upsample_mapping: np.ndarray) -> np.ndarray:
+    """Upsample using downsampled Sv and upsample mapping."""
+    upsampled_Sv = np.zeros_like(upsample_mapping, dtype=np.float64)
+    for upsampled_index, downsampled_index in enumerate(upsample_mapping):
+        upsampled_Sv[upsampled_index] = downsampled_Sv[downsampled_index]
+    return upsampled_Sv
+
+
+def downsample_upsample_along_depth(ds_Sv: xr.Dataset, depth_bin: float) -> xr.DataArray:
+    """
+    Downsample and upsample Sv to mimic what was done in echopy impulse
+    noise masking.
+    """
+    # Validate and compute range interval
+    depth_min = ds_Sv["depth"].min()
+    depth_max = ds_Sv["depth"].max()
+    range_interval = np.arange(depth_min, depth_max + depth_bin, depth_bin)
+    range_interval = _convert_bins_to_interval_index(range_interval)
+
+    # Downsample Sv along range sample
+    downsampled_Sv = flox.xarray.xarray_reduce(
+        ds_Sv["Sv"].pipe(_log2lin),
+        ds_Sv["channel"],
+        ds_Sv["ping_time"],
+        ds_Sv["depth"],
+        expected_groups=(None, None, range_interval),
+        isbin=[False, False, True],
+        method="map-reduce",
+        func="nanmean",
+        skipna=True,
+    ).pipe(_lin2log)
+
+    # Create upsample mapping
+    left_bin_values = [interval.left for interval in downsampled_Sv["depth_bins"].data]
+    upsample_mapping = xr.DataArray(
+        # Digitize denotes a value belonging in the first bin as being assigned to index 1 and prior
+        # to first bin as index 0. Since we want to create an index to index mapping between depth
+        # bins and the original Sv `depth`, this default behavior should be offset to the left,
+        # i.e a -1 applied to each value in the digitized array.
+        # Additionally, this subtraction will never result in -1 since `depth` will never contain
+        # any values prior to the first bin.
+        np.digitize(ds_Sv["depth"], left_bin_values) - 1,
+        dims=["channel", "ping_time", "range_sample"],
+    )
+
+    # Upsample the downsampled Sv
+    upsampled_Sv = xr.apply_ufunc(
+        _upsample_using_mapping,
+        # Need to compute arrays since indexing fails when delayed:
+        downsampled_Sv.compute(),
+        upsample_mapping.compute(),
+        input_core_dims=[["depth_bins"], ["range_sample"]],
+        output_core_dims=[["range_sample"]],
+        exclude_dims=set(["depth_bins", "range_sample"]),
+        dask="parallelized",
+        vectorize=True,
+        output_dtypes=[np.float64],
+    )
+    # Set range sample coords
+    upsampled_Sv = upsampled_Sv.assign_coords(range_sample=ds_Sv["range_sample"])
```

Comment:
I think you can use `xr.reindex_like` to achieve this, in the same way as what we talked about the other day, for the section in remove background noise to do forward fill after changing the coordinate of the downsampled array to the first index of each bin (which you have from `range_interval`).

Roughly (i have not tested these, but i am sure you get the gist):
```python
# Downsample Sv along range sample
downsampled_Sv = ...

# Substitute coordinate value in the downsampled Sv to be first element of each bin
downsampled_Sv.assign_coord({"depth": [FIRST_ELEMENT_OF_EACH_BIN]})

# Upsample via reindex_like
upsampled_Sv = downsampled_Sv.reindex_like(ds_Sv["Sv"], method="ffill")
```

Reply:
Ah yeah, I see what you mean. I can use the fact that depth is monotonic to `ffill`. I'll add that.

Reply:
Since we are assuming that depth and subsequently depth bins are not uniform, I had to do this `reindex` upsampling per channel, per ping time, in a nested for loop.

Reply:
I don't think the ffill approach will be a problem for non-uniform depths. After all it is jus the left bin edges (which is the first index you put into that `[FIRST_ELEMENT_OF_EACH_BIN]` list) that matters when you do ffill.

Reply:
oh i just realized what you mean -- it is when across ping_time the depth is different, then that becomes a problem. Ok, fair enough. But this is a weird messy situation, sigh. OK let's leave it as is, and just speed up in the assume-uniform (index only) case in a separate PR.

Reply:
Pretty unfortunate :/Should be much faster (and simpler) using reindex after any coarsen operation

Reply:
I wonder what your opinion is on if what I had initially is cleaner/better, with the ufunc and the index assignment mapping, or what I have now with reindex nested for-loop

Reply:
I am not sure... because I don't remember what the last one was. You can benchmark the performance? It would be good to know also.